File: backend/calculator/calculations/equity_intraday.py
from .base_calculator import BaseTradeCalculator
from .breakeven_calculator import BreakevenCalculator
from decimal import Decimal

class EquityIntradayCalculator(BaseTradeCalculator):
    def calculate_transaction_charges(self, transactions, position_type='long'):
        if self.platform != 'dhan':
            return {"error": "Intraday calculations are only implemented for Dhan broker at this time."}

        cumulative_quantity = Decimal("0")
        cumulative_buy_value = Decimal("0")
        total_buy_value = Decimal("0")
        total_sell_value = Decimal("0")
        total_brokerage = Decimal("0")
        transactions_data = []

        for transaction in transactions:
            quantity = Decimal(str(transaction["quantity"]))
            buy_price = Decimal(str(transaction["buyPrice"]))
            sell_price = Decimal(str(transaction["sellPrice"]))

            # For short positions, we need to interpret the inputs correctly
            if position_type == 'short':
                # In short positions, we sell first (at entry price) and buy later (at exit price)
                # Frontend sends: buyPrice = entry price (sell), sellPrice = exit price (buy)
                
                # Calculate values correctly from the input prices
                entry_price = buy_price     # Entry price (sell action)
                exit_price = sell_price     # Exit price (buy action)
                entry_value = quantity * entry_price  # Entry value (sell high)
                exit_value = quantity * exit_price    # Exit value (buy low)
                
                # For calculation consistency, map to buy/sell values
                sell_value = entry_value    # Sell value (entry action)
                buy_value = exit_value      # Buy value (exit action)
                
            else:
                # Long position - traditional buy then sell
                entry_price = buy_price
                exit_price = sell_price
                entry_value = quantity * entry_price
                exit_value = quantity * exit_price
                
                buy_value = entry_value
                sell_value = exit_value

            # Only include transactions with buy price > 0 in cumulative calculations for average
            if buy_price > 0:
                cumulative_quantity += quantity
                cumulative_buy_value += buy_value

            # Calculate brokerage per transaction (this should remain per-transaction)
            transaction_brokerage = self.broker.calculate_brokerage(buy_value, sell_value)
            total_brokerage += transaction_brokerage

            total_buy_value += buy_value
            total_sell_value += sell_value

            transactions_data.append(
                {
                    "quantity": str(quantity),
                    "buyValue": str(buy_value),
                    "sellValue": str(sell_value),
                    "averageBuyPrice": str(
                        (cumulative_buy_value / cumulative_quantity).quantize(Decimal("0.01"))
                        if cumulative_quantity > 0 and cumulative_buy_value > 0
                        else "0.00"
                    ),
                    "charges": str(transaction_brokerage.quantize(Decimal("0.01"))),
                }
            )

        # Calculate government charges on TOTAL/NET position (not per transaction)
        total_turnover = total_buy_value + total_sell_value
        stt = self.govt_charges.calculate_stt(total_sell_value)  # STT only on sell for intraday
        exchange_charges = self.govt_charges.calculate_exchange_charges(total_turnover)
        stamp_duty = self.govt_charges.calculate_stamp_duty(total_buy_value)
        sebi_fee = self.govt_charges.calculate_sebi_fee(total_turnover)
        ipft = self.govt_charges.calculate_ipft(total_turnover)

        # Calculate GST on total taxable components
        taxable_components = sum([total_brokerage, exchange_charges, sebi_fee, ipft])
        gst = self.govt_charges.calculate_gst(taxable_components)
        
        total_charges = sum([
            total_brokerage, stt, exchange_charges, stamp_duty, sebi_fee, ipft, gst
        ])
        # Calculate gross P&L based on position type
        if position_type == 'short':
            # For short positions: entry (sell) - exit (buy)
            # For short: profit = sell_value (entry) - buy_value (exit)
            gross_pnl = total_sell_value - total_buy_value
            
        else:
            # For long positions: exit (sell) - entry (buy)
            gross_pnl = total_sell_value - total_buy_value
            
        net_pnl = gross_pnl - total_charges
        
        # Calculate breakeven price
        breakeven_price = self.calculate_breakeven_price(cumulative_quantity, cumulative_buy_value, total_sell_value, total_charges, position_type, transactions)

        response_data = {
            "summary": {
                "totalQuantity": str(cumulative_quantity.quantize(Decimal("1"))),
                "totalBuyValue": str(total_buy_value.quantize(Decimal("0.01"))),
                "totalSellValue": str(total_sell_value.quantize(Decimal("0.01"))),
                "averageBuyPrice": str(
                    (cumulative_buy_value / cumulative_quantity).quantize(Decimal("0.01"))
                    if cumulative_quantity > 0 and cumulative_buy_value > 0
                    else "0.00"
                ),
                "turnover": str(total_turnover.quantize(Decimal("0.01"))),
                "grossPnL": str(gross_pnl.quantize(Decimal("0.01"))),
                "netPnL": str(net_pnl.quantize(Decimal("0.01"))),
                "breakevenPrice": str(breakeven_price.quantize(Decimal("0.01"))),
            },
            "charges": {
                "brokerage": str(total_brokerage.quantize(Decimal("0.01"))),
                "stt": str(stt.quantize(Decimal("0.01"))),
                "exchangeCharges": str(exchange_charges.quantize(Decimal("0.01"))),
                "stampDuty": str(stamp_duty.quantize(Decimal("0.01"))),
                "sebiFee": str(sebi_fee.quantize(Decimal("0.01"))),
                "ipft": str(ipft.quantize(Decimal("0.01"))),
                "gst": str(gst.quantize(Decimal("0.01"))),
                "dpCharges": str(Decimal('0.00')),
                "totalCharges": str(total_charges.quantize(Decimal("0.01"))),
            },
            "transactions": transactions_data,
        }
        return response_data 
        
    def calculate_breakeven_price(self, quantity, buy_value, sell_value, total_charges, position_type='long', transactions=None):
        """
        Calculate the breakeven price for a position.
        For long positions: The minimum exit (sell) price to avoid loss.
        For short positions: The maximum exit (buy) price to avoid loss.
        """
        if quantity <= 0:
            return Decimal("0")
        
        # Calculate entry price per share based on position type
        if position_type == 'long':
            # For long positions, entry is buy
            entry_price = buy_value / quantity if quantity > 0 else Decimal("0")
        else:
            # For short positions, entry is the buyPrice from frontend (sell action)
            # Frontend sends: buyPrice = entry price (sell), sellPrice = exit price (buy)
            if transactions and len(transactions) > 0:
                entry_price = Decimal(str(transactions[0]["buyPrice"]))
            else:
                entry_price = sell_value / quantity if quantity > 0 else Decimal("0")
        
        # Use the binary search method for more accurate breakeven calculation
        return BreakevenCalculator.calculate_breakeven_price(
            quantity=quantity,
            entry_price=entry_price,
            broker=self.broker,
            exchange=self.exchange,
            trade_type=self.trade_type,
            position_type=position_type
        )

File: backend-node/src/calculations/equityIntraday.ts
import { BaseTradeCalculator, DecimalUtils } from './index.js';
import Decimal from 'decimal.js';
import { BreakevenCalculator } from './breakevenCalculator.js';

export class EquityIntradayCalculator extends BaseTradeCalculator {
  calculate_transaction_charges(transactions: any[], positionType: string = 'long') {
    if (this.platform !== 'dhan') {
      return { error: "Intraday calculations are only implemented for Dhan broker at this time." };
    }

    let cumulative_quantity = DecimalUtils.zero();
    let cumulative_buy_value = DecimalUtils.zero();
    let total_buy_value = DecimalUtils.zero();
    let total_sell_value = DecimalUtils.zero();
    let total_brokerage = DecimalUtils.zero();
    let transactions_data: any[] = [];

    for (const transaction of transactions) {
      const quantity = DecimalUtils.parse(transaction.quantity);
      const buy_price = DecimalUtils.parse(transaction.buyPrice);
      const sell_price = DecimalUtils.parse(transaction.sellPrice);

      // For short positions, we need to interpret inputs correctly
      let buy_value: Decimal;
      let sell_value: Decimal;
      
      if (positionType === 'short') {
        // In short positions: entry price (sell) -> buyPrice, exit price (buy) -> sellPrice
        const entry_price = buy_price;     // Entry price (sell action)
        const exit_price = sell_price;     // Exit price (buy action)
        
        buy_value = quantity.mul(exit_price);  // Buy to close
        sell_value = quantity.mul(entry_price); // Sold to open
      } else {
        // Long position - traditional buy then sell
        buy_value = quantity.mul(buy_price);
        sell_value = quantity.mul(sell_price);
      }

      // Only include transactions with buy price > 0 in cumulative calculations for average
      if (buy_price.greaterThan(DecimalUtils.zero())) {
        cumulative_quantity = cumulative_quantity.add(quantity);
        cumulative_buy_value = cumulative_buy_value.add(buy_value);
      }

      // Calculate brokerage per transaction
      const transaction_brokerage = this.broker.calculate_brokerage(buy_value, sell_value);
      total_brokerage = total_brokerage.add(transaction_brokerage);

      total_buy_value = total_buy_value.add(buy_value);
      total_sell_value = total_sell_value.add(sell_value);

      const buyValueString = positionType === 'short'
        ? DecimalUtils.multiplyToPythonString(transaction.quantity, transaction.sellPrice, buy_value)
        : DecimalUtils.multiplyToPythonString(transaction.quantity, transaction.buyPrice, buy_value);

      const sellValueString = positionType === 'short'
        ? DecimalUtils.multiplyToPythonString(transaction.quantity, transaction.buyPrice, sell_value)
        : DecimalUtils.multiplyToPythonString(transaction.quantity, transaction.sellPrice, sell_value);

      transactions_data.push({
        quantity: DecimalUtils.toPythonStringFromInput(transaction.quantity, quantity),
        buyValue: buyValueString,
        sellValue: sellValueString,
        averageBuyPrice: cumulative_quantity.greaterThan(DecimalUtils.zero()) && cumulative_buy_value.greaterThan(DecimalUtils.zero())
          ? DecimalUtils.quantizeToString(cumulative_buy_value.div(cumulative_quantity), 2)
          : '0.00',
        charges: DecimalUtils.quantizeToString(transaction_brokerage, 2)
      });
    }

    // Calculate government charges on TOTAL/NET position
    const total_turnover = total_buy_value.add(total_sell_value);
    const stt = this.govtCharges.calculate_stt(total_sell_value); // STT only on sell for intraday
    const exchange_charges = this.govtCharges.calculate_exchange_charges(total_turnover);
    const stamp_duty = this.govtCharges.calculate_stamp_duty(total_buy_value);
    const sebi_fee = this.govtCharges.calculate_sebi_fee(total_turnover);
    const ipft = this.govtCharges.calculate_ipft(total_turnover);

    // Calculate GST on total taxable components
    const taxable_components = total_brokerage.add(exchange_charges).add(sebi_fee).add(ipft);
    const gst = this.govtCharges.calculate_gst(taxable_components);
    
    const total_charges = total_brokerage
      .add(stt)
      .add(exchange_charges)
      .add(stamp_duty)
      .add(sebi_fee)
      .add(ipft)
      .add(gst);

    // Calculate gross P&L based on position type
    const gross_pnl = total_sell_value.sub(total_buy_value);
    const net_pnl = gross_pnl.sub(total_charges);
    
    // Django-parity breakeven
    const breakeven_price = this.calculate_breakeven_price(positionType, transactions, cumulative_quantity, cumulative_buy_value, total_sell_value);

    return {
      summary: {
        totalQuantity: DecimalUtils.quantizeToString(cumulative_quantity, 0),
        totalBuyValue: DecimalUtils.quantizeToString(total_buy_value, 2),
        totalSellValue: DecimalUtils.quantizeToString(total_sell_value, 2),
        averageBuyPrice: cumulative_quantity.greaterThan(DecimalUtils.zero()) && cumulative_buy_value.greaterThan(DecimalUtils.zero())
          ? DecimalUtils.quantizeToString(cumulative_buy_value.div(cumulative_quantity), 2)
          : '0.00',
        turnover: DecimalUtils.quantizeToString(total_turnover, 2),
        grossPnL: DecimalUtils.quantizeToString(gross_pnl, 2),
        netPnL: DecimalUtils.quantizeToString(net_pnl, 2),
        breakevenPrice: DecimalUtils.quantizeToString(breakeven_price, 2)
      },
      charges: {
        brokerage: DecimalUtils.quantizeToString(total_brokerage, 2),
        stt: DecimalUtils.quantizeToString(stt, 2),
        exchangeCharges: DecimalUtils.quantizeToString(exchange_charges, 2),
        stampDuty: DecimalUtils.quantizeToString(stamp_duty, 2),
        sebiFee: DecimalUtils.quantizeToString(sebi_fee, 2),
        ipft: DecimalUtils.quantizeToString(ipft, 2),
        gst: DecimalUtils.quantizeToString(gst, 2),
        dpCharges: '0.00', // No DP charges for intraday
        totalCharges: DecimalUtils.quantizeToString(total_charges, 2)
      },
      transactions: transactions_data
    };
  }

  private calculate_breakeven_price(
    position_type: string,
    transactions: any[],
    quantity: Decimal,
    cumulative_buy_value: Decimal,
    total_sell_value: Decimal
  ): Decimal {
    if (quantity.lessThanOrEqualTo(DecimalUtils.zero())) {
      return DecimalUtils.zero();
    }

    let entry_price: Decimal;
    if (position_type === 'long') {
      entry_price = cumulative_buy_value.greaterThan(DecimalUtils.zero())
        ? cumulative_buy_value.div(quantity)
        : DecimalUtils.zero();
    } else {
      // Django: for short, entry is the `buyPrice` field from the frontend payload.
      if (transactions && transactions.length > 0) {
        entry_price = DecimalUtils.parse(transactions[0].buyPrice);
      } else {
        entry_price = total_sell_value.greaterThan(DecimalUtils.zero())
          ? total_sell_value.div(quantity)
          : DecimalUtils.zero();
      }
    }

    return BreakevenCalculator.calculate_breakeven_price(
      quantity,
      entry_price,
      this.broker,
      this.exchange,
      this.tradeType,
      position_type
    );
  }
}
